drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,765 @@
+ */
+static int i2c_omap_set_speed(const struct device *dev, uint32_t speed)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+
+	/* If configured for High Speed */
+	switch (speed) {
+	case I2C_BITRATE_FAST:
+		/* Fast mode */
+		data->speed_config = (struct i2c_omap_speed_config){
+			.pscstate = 9,
+			.scllstate = 7,
+			.sclhstate = 5,
+		};
+		break;
+	case I2C_BITRATE_STANDARD:
+		/* Standard mode */
+		data->speed_config = (struct i2c_omap_speed_config){
+			.pscstate = 23,
+			.scllstate = 13,
+			.sclhstate = 15,
+		};
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Configure the OMAP I2C controller with the specified device configuration.
+ *
+ * This function configures the OMAP I2C controller with the specified device configuration.
+ *
+ * @param dev The pointer to the device structure.
+ * @param dev_config The device configuration to be applied.
+ *
+ * @return 0 on success, negative error code on failure.
+ */
+static int i2c_omap_configure(const struct device *dev, uint32_t dev_config)
+{
+	uint32_t speed_cfg;
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		speed_cfg = I2C_BITRATE_STANDARD;
+	case I2C_SPEED_FAST:
+		speed_cfg = I2C_BITRATE_FAST;
+	default:
+		return -ENOTSUP;
+	}
+	if ((dev_config & I2C_MODE_CONTROLLER) != I2C_MODE_CONTROLLER) {
+		return -ENOTSUP;
+	}
+	if ((dev_config & I2C_MSG_ADDR_10_BITS) != I2C_MSG_ADDR_10_BITS) {
+		return -ENOTSUP;
+	}
+	i2c_base_addr->CON = 0;
+	i2c_omap_set_speed(dev, speed_cfg);
+	i2c_omap_init_ll(dev);
+	return 0;
+}
+
+/**
+ * @brief Transmit or receive data over I2C bus
+ *
+ * This function transmits or receives data over the I2C bus using the OMAP I2C controller.
+ *
+ * @param dev Pointer to the I2C device structure
+ * @param num_bytes Number of bytes to transmit or receive
+ */
+static void i2c_omap_transmit_receive_data(const struct device *dev, uint8_t num_bytes)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint8_t *buf_ptr = data->current_msg.buf;
+
+	while (num_bytes--) {
+		if (data->receiver) {
+			*buf_ptr++ = i2c_base_addr->DATA;
+		} else {
+			i2c_base_addr->DATA = *(buf_ptr++);
+		}
+	}
+}
+
+/**
+ * @brief Resize the FIFO buffer for the OMAP I2C controller.
+ *
+ * This function resizes the FIFO buffer for the OMAP I2C controller based on the specified size.
+ * It clears the RX threshold and sets the new size for the receiver, or clears the TX threshold
+ * and sets the new size for the transmitter.
+ *
+ * @param dev Pointer to the device structure.
+ * @param size The new size of the FIFO buffer.
+ */
+static void i2c_omap_resize_fifo(const struct device *dev, uint8_t size)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
+	if (data->receiver) {
+		i2c_base_addr->BUF |= ((size) << 8) | I2C_OMAP_BUF_RXFIF_CLR;
```

Comment:
You only seem to be OR'ing into ->BUF, how do these bits get cleared? I think you are missing some masking.

Reply:
The mask was Rx/Tx DMA channel Enable. As we aren't using it as of now. I didn't mask it out.